pdfrw/objects/pdfstring.py

```diff
@@ -1,74 +1,507 @@
+        return convert_store(''.join(result))
+
+
+    def decode_hex(self):
+        """ Decode a PDF hexadecimal-encoded string, which is enclosed
+            in angle brackets <>.
+        """
+        hexstr = convert_store(''.join(self[1:-1].split()))
+        if len(hexstr) % 1: # odd number of chars indicates a truncated 0
+            hexstr += '0'
+        return binascii.unhexlify(hexstr)
+
+
+    def decode(self, to_bytestring=False):
+        """ Decode a PDF string.  This is a convenience function
+            for user code, in that (as of pdfrw 0.3) it is never
+            actually used inside pdfrw.
+
+            A PDF string might represent Unicode code points, or
+            it might just be a collection of raw bytes.  That
+            distinction is context-dependent, so it is up to the
+            caller of this function to set to_bytestring to True
+            if it does not want a Unicode string result.
+
+            There are two Unicode storage methods used -- either
+            UTF16_BE, or something called PDFDocEncoding, which
+            is defined in the PDF spec.  The determination of
+            which decoding method to use is done by examining the
+            first two bytes for the byte order marker.
+        """
+        if self.startswith('(') and self.endswith(')'):
+            raw = self.decode_literal()
+
+        elif self.startswith('<') and self.endswith('>'):
+            raw = self.decode_hex()
 
         else:
-            return self.decode_hex(remap, twobytes)
-
-    def encode(cls, source, usehex=False):
-        assert not usehex, "Not supported yet"
-        source = source.replace('\\', '\\\\')
-        source = source.replace('(', '\\(')
-        source = source.replace(')', '\\)')
-        return cls('(' + source + ')')
-    encode = classmethod(encode)
+            raise ValueError('Invalid PDF string "%s"' % repr(self))
+
+        if to_bytestring:
+            return raw
+        elif raw[:2] == codecs.BOM_UTF16_BE:
+            return raw[2:].decode('utf-16-be')
+        else:
+            return raw.decode('pdfdocencoding')
+
+    # These constants are used for the mode parameter to encode()
+
+    ENCODE_PDFDOC = 1   # Use PDFDocEncoding to encode Unicode
+    ENCODE_UTF16 = 2    # Use UTF-16BE to encode Unicode
+    ENCODE_LITERAL = 4  # Use () literal strings for byte encoding
+    ENCODE_HEX = 8      # Use <> hex strings for byte encoding
+
+    # Internal values used by encode
+
+    bytes_bom = codecs.BOM_UTF16_BE
+    uni_bom = codecs.BOM_UTF16_BE.decode('latin-1')
+    uni_type = type(uni_bom)
+    escape_splitter = None  # Calculated on first use
+
+    @classmethod
+    def init_escapes(cls):
+        """ Initialize the escape_splitter for the encode method
+        """
+        cls.escape_splitter = re.compile(br'(\(|\\|\))').split
+        return cls.escape_splitter
+
+    @classmethod
+    def encode(cls, source, mode=0):
```

Comment:
In my experience, an integer parameter used as a bit field is usually called `flags` rather than `mode`.

I might lean towards a more high-level API, though:

- `text_encoding='pdfdocencoding' or 'utf16' or 'auto'`
- `bytes_encoding='literal' or 'hex' or 'auto'`